Add callback function for the inserter

prnvbn opened the pull request:
## Summary
Addresses https://github.com/ClickHouse/clickhouse-rs/issues/306

## Checklist
Delete items not relevant to your PR:
- [ ] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided so that we can include it in CHANGELOG later

abonander commented on src/inserter.rs:
```diff
@@ -32,6 +32,8 @@ pub struct Inserter<T> {
     pending: Quantities,
     in_transaction: bool,
+
+    on_commit: Option<Box<dyn FnMut(&Quantities) + Send + 'static>>,
```
I think keeping it un-aliased is actually better for readability as long as the signature isn't too complicated.

abonander commented on src/inserter.rs:
```diff
@@ -199,6 +202,16 @@ where
+    /// The callback receives the committed [`Quantities`]. It is invoked only
+    /// when a batch actually commits (i.e., non-zero transactions), and only
+    /// after the commit completes successfully.
+    pub fn with_on_commit(mut self, callback: impl FnMut(&Quantities) + Send + 'static) -> Self {
```
Hmm, `with_on_commit` is a little clunky but the `with_` prefix does match the other combinators.

Maybe `with_commit_callback` or just `on_commit`? I don't feel particularly strongly about any of these, though.

abonander commented on src/inserter.rs:
```diff
@@ -264,6 +277,12 @@ where
         self.ticks.reschedule();
         result?;
+
+        if let Some(cb) = &mut self.on_commit {
+            if quantities.transactions > 0 {
+                (cb)(&quantities);
+            }
+        }
```
You could lift this into `insert()` along with `mem::replace(&mut self.pending, Quanities::ZERO)` and the behavior should still be correct.

prnvbn replied:
addressed in [ee6b162](https://github.com/ClickHouse/clickhouse-rs/pull/307/commits/ee6b162e56a6e879c26c1f06938abea670ee1538)

abonander reviewed (changes requested):
A couple of nits, otherwise looks fine.

prnvbn commented:
@abonander addressed the comments


(P.S. I merged instead of a rebase as I can see from the commit history that this PR will be squashed and merged)

slvrtrn commented:
There are some failures with rustfmt and clippy; otherwise, it looks good

prnvbn commented:
> There are some failures with rustfmt and clippy; otherwise, it looks good

addressed most of them. suppressed one as per - https://github.com/ClickHouse/clickhouse-rs/pull/307#discussion_r2392770663

prnvbn commented:
bump on this

abonander reviewed (approved):
This looks good to me, but it's an open question whether we want to block this on adding a test.

I'm personally on the fence about it since the added functionality is rather trivial, so as long as inserts continue to function correctly (which is covered already) it doesn't seem like it needs a specific test. 

But I'd like @slvrtrn's opinion on it before merging.

slvrtrn commented:
Let's add more tests as a follow-up.
